flink-tests/src/test/java/org/apache/flink/test/scheduling/SpeculativeExecutionITCase.java

```diff
@@ -524,30 +542,66 @@ public void open(GenericInputSplit split) throws IOException {
     }
 
     private static class TestingNumberSequenceSource extends NumberSequenceSource {
-        private TestingNumberSequenceSource() {
+
+        private final boolean forceFailureFlag;
+
+        private TestingNumberSequenceSource(boolean forceFailureFlag) {
             super(0, NUMBERS_TO_PRODUCE - 1);
+            this.forceFailureFlag = forceFailureFlag;
         }
 
         @Override
         public SourceReader<Long, NumberSequenceSplit> createReader(
                 SourceReaderContext readerContext) {
-            return new TestingIteratorSourceReader(readerContext);
+            return new TestingIteratorSourceReader(readerContext, forceFailureFlag);
+        }
+
+        @Override
+        public SplitEnumerator<NumberSequenceSplit, Collection<NumberSequenceSplit>>
+                createEnumerator(final SplitEnumeratorContext<NumberSequenceSplit> enumContext) {
+
+            int splitSize = enumContext.currentParallelism();
+            if (forceFailureFlag) {
+                splitSize = 1;
```

Comment:
Not sure why it should be 1 instead of 0(no split)?

And comments are needed to explain such kind of magic.

Reply:
I have add a comment to explain it.

Reply:
if 0 split, the source will not produce any data, leading to a mismatch in unit test results.